core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala

```diff
@@ -282,6 +317,188 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
+          val nextAttempts =
+            taskScheduler.resourceOffers(IndexedSeq(WorkerOffer("executor4", "host4", 1))).flatten
+          if (attempt < 3) {
+            assert(nextAttempts.size === 1)
+            task = nextAttempts(0)
+            assert(task.index === taskIndex)
+          } else {
+            assert(nextAttempts.size === 0)
+          }
+        }
+        // End the other task of the taskset, doesn't matter whether it succeeds or fails.
+        val otherTask = tasks(1)
+        val result = new DirectTaskResult[Int](valueSer.serialize(otherTask.taskId), Seq())
+        tsm.handleSuccessfulTask(otherTask.taskId, result)
+      } else {
+        tasks.foreach { task =>
+          val result = new DirectTaskResult[Int](valueSer.serialize(task.taskId), Seq())
+          tsm.handleSuccessfulTask(task.taskId, result)
+        }
+      }
+      assert(tsm.isZombie)
+    }
+  }
+
+  /**
+   * Helper for performance tests.  Takes the explicitly blacklisted nodes and executors; verifies
+   * that the blacklists are used efficiently to ensure scheduling is not O(numPendingTasks).
+   */
+  private def blacklistPerformanceCheck(
+      nodeBlacklist: Seq[String],
+      execBlacklist: Seq[String]): Unit = {
+    // When an executor or node is blacklisted, we want to make sure that we don't try scheduling
+    // each pending task, one by one, to discover they are all blacklisted.  This is important for
+    // performance -- if we did check each task one-by-one, then responding to a resource offer
+    // (which is usually O(1)-ish) would become O(numPendingTasks), which would slow down
+    // scheduler throughput and slow down scheduling even on healthy executors.
+    // Here, we check a proxy for the runtime -- we make sure the scheduling is short-circuited
+    // at the node or executor blacklist, so we never check the per-task blacklist.  We also make
+    // sure we don't check the node & executor blacklist for the entire taskset O(numPendingTasks)
+    // times.
+
+    taskScheduler = setupSchedulerWithMockTaskSetBlacklist()
+    // we schedule 500 tasks so we can clearly distinguish anything that is O(numPendingTasks)
+    val taskSet = FakeTask.createTaskSet(numTasks = 500, stageId = 0, stageAttemptId = 0)
+    taskScheduler.submitTasks(taskSet)
+
+    val offers = IndexedSeq(
+      new WorkerOffer("executor1", "host1", 1),
+      new WorkerOffer("executor2", "host1", 1),
+      new WorkerOffer("executor3", "host2", 1)
+    )
+    // We should check the node & exec blacklists, but only O(numOffers), not O(numPendingTasks)
+    // times.  Its O(numOffers), not exactly numOffers, because we offer resources multiple
+    // times.  An upper bound on the worst case is -- we check it once for every core at each
+    // locality level.  (We could tighten the bound a bit but that should be a good enough check.)
+    val numCoresOnAllOffers = offers.map(_.cores).sum
```

Comment:
In this particular case, we might be able to tighten the bound.  But I was also trying to right this in a way that it didn't involve too many magic constants, if somebody modified it in the future.  Imagine there were 500 cores offered, but after shuffling the offers around, the blacklisted once was first in the list of offers.  Then when you call [`resourceOfferSingleTaskSet`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L257), you'd need to get through the blacklisted resource before scheduling something.  But since you *do* launch something on the offers later in the shuffled list, you'd [call it again](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L330), and every iteration, you repeat that blacklist check.

this could be improved a lot by either:
(a) pulling the filtering from the taskset blacklist further out of those nested loops, out to [`taskScheduler.resourceOffer`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L328)
or (b) push that entire `do...while` down into `TaskSetManager`, so that the TaskSetManager is responsible for more assigning more than one task at the time.
(the same logic either way, just a question of code organization.)

Though `O(numOffers)` sounds bad, I decided it was OK because most of the time this is called with one offer (the resource which just freed up from a completed task).  And when you do have a ton of offers, eg. when you start a stage, well scheduling is overall `O(numOffers)` anyhow.

Thinking about this again, pretty sure the bound can actually be `numCores + numLocalityLevels` -- each iteration either schedules something, or it terminates that locality level.